paasta_tools/paasta_metastatus.py

```diff
@@ -338,9 +349,50 @@ def main(argv: Optional[List[str]]=None) -> None:
     metastatus_lib.print_results_for_healthchecks(chronos_summary, chronos_ok, chronos_results, args.verbose)
 
     if not healthy_exit:
-        sys.exit(2)
-    else:
-        sys.exit(0)
+        raise FatalError(2)
+
+
+PAASTA_PRINT_PATCH_LOCK = Lock()
+
+
+def get_output(argv: Optional[List[str]]=None) -> Tuple[str, int]:
+    output = io.StringIO()
+
+    def patched_paasta_print(*args, **kwargs):
+        if 'file' in kwargs:
+            del kwargs['file']
+        return print(*args, **kwargs, file=output)
+
+    exit_code = 1
+    PAASTA_PRINT_PATCH_LOCK.acquire()
+    global paasta_print
+    orig_paasta_print = paasta_print
+    try:
+        # Monkey-patching `paasta_print` everywhere it's used to output results of `paasta metastatus` cmd.
```

Comment:
i'm not a fan of such patching, eventually we will shoot ourselves in the foot. capturing output is tricky and there's definitely more than one way we print stuff in paasta.

ideally we would rewrite `print_output` to simply do `paasta_print(data for data in get_output())` and make a `get_output` function  that returns list of messages (or a generator) instead of printing them.

i realize that this may be a lot of changes, but i think it's important.

Reply:
The current design is a compromise to make "`paasta metastatus` via HTTP" change as simple as possible.

Ideally we probably should export all `metastatus_lib` functions we use to generate `paasta metastatus` output via HTTP API and use them on the client.  I've tried to start with that, but it seems that it's a way too much work for the result.

What do you think about introducing a context manager which sets output destination for `paasta_print`, and change `paasta_print` to output everything into that destination if it's set?  This way we will not need any monkey patching?
I.e.
```
with set_paasta_print_file(output):
    print_output(argv)
```

Reply:
context looks better, though it's still going to try to capture output behind the scenes. i don't want to block progress, so ship away, but in future we need to rethink this.

Reply:
Replaced with contextmanager-based code.